File: Java/1390. Four Divisors.java
class Solution {
    public int sumFourDivisors(int[] nums) {
        int ans = 0;
        Map<Integer, Integer> divisors = new HashMap<>();

        for (int num: nums) {
            if (divisors.containsKey(num)) {
                ans += divisors.get(num);
                continue;
            }
            ans += findDivisors(divisors, num);
        }

        return ans;
    }

    private int findDivisors(Map<Integer, Integer> divisors, int num) {
        int count = 0, sum = 0;
        for (int d = 1; d <= Math.floor(Math.sqrt(num)); d++) {
            if ((num / d) * d == num) {
                if (num / d != d) {
                    count++;
                    sum += num / d;
                }
                count++;
                sum += d;
            }

            if (count > 4) {
                break;
            }
        }
        
        if (count == 4) {
            divisors.put(num, sum);
        }

        return count == 4 ? sum : 0;
    }
}

File: Python3/1390. Four Divisors.py
class Solution:
    def sumFourDivisors(self, nums: List[int]) -> int:
        divisors = defaultdict(int)
        ans = 0

        def findDivisors(n: int) -> int:
            if n in divisors:
                return divisors[n]

            count = _sum = 0
            for d in range(1, int(sqrt(n)) + 1):
                if n / d == n // d:
                    if d != n // d:
                        count += 1
                        _sum += n // d
                    count += 1
                    _sum += d

                if count > 4:
                    break

            if count == 4:
                divisors[n] = _sum
                return _sum
            else:
                return 0

        for n in nums:
            ans += findDivisors(n)

        return ans